### ISSUE TITLE
Missing extern token causes weird runtime behavior

### ISSUE THREAD

**Issue by arnaudgolfouse** [lalrpop/lalrpop#960]
Here is a simplified version of a project a friend of mine is working on:

<details> 
  <summary>Project files</summary>

`main.rs`
```rust
pub mod ast;
pub mod lexer;

use lalrpop_util::lalrpop_mod;
lalrpop_mod!(grammar);

use lexer::Lexer;

fn main() {
    let input = "1";
    let res = grammar::LiteralParser::new()
        .parse(Lexer::new(input))
        .unwrap();
    println!("{:?}", res);
}
```

`lexer.rs`
```rust
use logos::Logos;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Logos)]
#[logos(error = LexerError)]
#[logos(skip r"[ \t\n\f]+")] // Ignore this regex pattern between tokens
pub enum Token<'input> {
    #[token("=")]
    Assign,
    #[token("let")]
    KeywordLet,
    #[regex("[0-9]+", |lex| lex.slice().parse().ok())]
    Integer(i64),
    #[regex("[a-zA-Z]+")]
    Identifier(&'input str),
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
pub struct LexerError;

pub struct Lexer<'input> {
    inner: logos::SpannedIter<'input, Token<'input>>,
}
impl<'input> Lexer<'input> {
    pub fn new(input: &'input str) -> Self {
        Self {
            inner: Token::lexer(input).spanned(),
        }
    }
}
impl<'input> Iterator for Lexer<'input> {
    type Item = Result<(usize, Token<'input>, usize), LexerError>;
    fn next(&mut self) -> Option<Self::Item> {
        let (next_token, range) = self.inner.next()?;
        Some(next_token.map(|t| (range.start, t, range.end)))
    }
}
```

`ast.rs`
```rust
use crate::lexer::Token;

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum Ast<'input> {
    LetBinding(&'input str, Box<Self>),
    Literal(Token<'input>),
}
```

`grammar.lalrpop`
```rust
use crate::lexer::{Token, LexerError};
use crate::ast::Ast;
use Token::*;

grammar<'input>;

extern {
    type Location = usize;
    type Error = LexerError;

    enum Token<'input> {
        "=" => Assign,
        "let" => KeywordLet,
        "in" => KeywordIn,
        "int" => Integer(<i64>),
        "IDENT" => Identifier(<&'input str>),
    }
}

pub LetBinding : Ast<'input> = {
    "let" <i : "IDENT">  "=" <e : Expr> => Ast::LetBinding(i, Box::new(e)),
};

pub Expr : Ast<'input> = {
    <l:Literal> => l,
};

pub Literal : Ast<'input> = {
    <i : "int"> => Ast::Literal(Token::Integer(i)),
};
```

`build.rs`
```rust
fn main() { lalrpop::process_root().unwrap() }
```

</details>

Now when running with `cargo run`, it crashes with an error!
```
called `Result::unwrap()` on an `Err` value: UnrecognizedToken { token: (0, Integer(1), 1), expected: ["\"int\""] }
```

What a weird error 🤔 What happened is, we forgot to declare `KeywordIn` in `lexer.rs`! Putting it back gives us the expected behavior.

Now of course the initial example was quite bigger, and it was a bit of a headache to understand what was going on.

It would be nice if lalrpop could somehow give an error in this case.

(lalrpop version is `0.21.0`, logos version is `0.14.2`)

**Comment by dburgener** [comment]
Poking at this a bit, in the generated code from the above, we generate the following match statement:

```
        match __token {
            Assign if true => Some(0),
            KeywordLet if true => Some(1),
            KeywordIn if true => Some(2),
            Integer(_) if true => Some(3),
            Identifier(_) if true => Some(4),
            _ => None,
        }
```

where `__token` is a `Token` as defined above.  What's interesting about this match statement is that KeywordIn isn't a match variant.  So rust interprets it as a *variable name* instead.  That means that any token that fails to match Assign or KeywordLet ends up binding to the variable KeywordIn and skipping future match arms.

We suppress a lot of warnings in the generated grammar (for good reason!) but if we turn those off, we get helpful warnings around "variable KeywordIn should have a snake case name" and "unused variable KeywordIn".  If those warnings showed up by themselves, that could probably tip a user off that KeywordIn is being interpreted as a variable, instead of the intent of interpreting it as an enum variant.  But we are suppressing a lot of other snake case/unused variables in the generated code by design, and the warnings are noisy.  Maybe future work could be to try for more targeted suppressions to avoid suppressing helpful warnings in cases like this.

It would be nice if lalrpop could detect which enum variants are valid and generate its match with only valid variants.  I don't think that's possible though (or at least not easy).  lexer.rs hasn't been compiled yet at build time when we output the generated parser, and so learning what variants are valid at that point is highly nontrivial.

A change to the user code like this:

```
diff --git a/src/grammar.lalrpop b/src/grammar.lalrpop
index 9d264ef..da498b6 100644
--- a/src/grammar.lalrpop
+++ b/src/grammar.lalrpop
@@ -1,6 +1,5 @@
 use crate::lexer::{Token, LexerError};
 use crate::ast::Ast;
-use Token::*;
 
 grammar<'input>;
 
@@ -9,11 +8,11 @@ extern {
     type Error = LexerError;
 
     enum Token<'input> {
-        "=" => Assign,
-        "let" => KeywordLet,
-        "in" => KeywordIn,
-        "int" => Integer(<i64>),
-        "IDENT" => Identifier(<&'input str>),
+        "=" => Token::Assign,
+        "let" => Token::KeywordLet,
+        "in" => Token::KeywordIn,
+        "int" => Token::Integer(<i64>),
+        "IDENT" => Token::Identifier(<&'input str>),
     }
 }
```

to avoid the glob import, and make explicit what the enum variants are results in a match arm that can't be misinterpreted and compiler errors like:

```
error[E0599]: no variant or associated item named `KeywordIn` found for enum `Token` in the current scope
   --> /home/daniel/issue_960/target/debug/build/issue_960-c81999f0075ecefd/out/grammar.rs:204:20
    |
204 |             Token::KeywordIn if true => Some(2),
    |                    ^^^^^^^^^
    |                    |
    |                    variant or associated item not found in `Token<'_>`
    |                    help: there is a variant with a similar name: `KeywordLet`
    |
   ::: src/lexer.rs:6:1
    |
6   | pub enum Token<'input> {
    | ---------------------- variant or associated item `KeywordIn` not found for this enum
```

Which I think should be sufficient for a user to diagnose the problem (at a minimum, better than the error reported here).

The logos [example](https://lalrpop.github.io/lalrpop/lexer_tutorial/005_external_lib.html#update-the-grammar) in the lalrpop book, uses explicit enum variants rather than the glob import, so we are providing good guidance to users in that sense.

I agree that the situation is suboptimal, but I'm not sure there's a lot we can do about it.  You can get this exact sort of situation in pure rust by glob importing your enum pattern, and writing a match which misinterprets a missing variant as a variable.  Lalrpop as a middle man can't tell what the valid variants are to warn that way.  And it doesn't know the enum name to insert the variant name itself.  At most maybe it could warn about the the glob import as an anti-pattern, but there are likely legitimate usecases for it (eg importing all times from an ast file), so that seems too noisy.

**Comment by arnaudgolfouse** [comment]
Ah I see, thanks for the explanation!
Indeed I think there isn't much lalrpop can do in this case... maybe putting `#[allow(non_snake_case)]` on top of this match statement could be possible, if all generated variables are snake_case? I do understand that the gains are dubious though.

As a side note, there _were_ warnings related to the glob import (unused import: `Token::*`), we just didn't interpret them correctly 😅

**Comment by dburgener** [comment]
>  maybe putting #[allow(non_snake_case)] on top of this match statement could be possible, if all generated variables are snake_case?

Do you mean `#[warn(non_snake_case)]`?  It's already in a block with `#[allow(non_snake_case)]`, which is why the warning didn't fire.  I just tested manually adding it on the match in this case and it fired here and seems like it would be helpful at least for this case.  That might indeed be a reasonable thing to add here.  I'll think on it some more and maybe submit a PR for it later today.

**Comment by dburgener** [comment]
Opened #961

**Comment by arnaudgolfouse** [comment]
> > maybe putting #[allow(non_snake_case)] on top of this match statement could be possible, if all generated variables are snake_case?
> 
> Do you mean `#[warn(non_snake_case)]`? It's already in a block with `#[allow(non_snake_case)]`, which is why the warning didn't fire. I just tested manually adding it on the match in this case and it fired here and seems like it would be helpful at least for this case. That might indeed be a reasonable thing to add here. I'll think on it some more and maybe submit a PR for it later today.

Yeah that's what I meant 😁 thanks for the quick response and for the fix, much appreciated!
